### PR TITLE
Update version MQTTnet library

### PR DESCRIPTION
**Pull request by osmanhadzic**
Update version MQTTnet library from 3.1.2 to 4.1.4.563. Refactor strategy for fallback messages.

Updating the MQTTnet library provides benefits such as bug fixes, security patches, performance improvements, new features, compatibility with the latest .NET versions, and better community support.

Implement breaking changes:
 - Rename namespaces 
 - Implement an enum for Qos 
 - Replace handler with events (messages receive and disconnect)

Handling exception to resolve deadlock in threads. With add cancellation timeout token in class `AstarteMqttTransport` method `Connect()`  and handling `AggregateException` in class `AstarteDevice` method `EventualyReconnect()`.

### CODE REVIEW

File: AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs [hunk 1]
```diff
@@ -114,18 +121,30 @@ public override async Task Connect()
...
+                        ($"Error connecting to MQTT. Code: {result.ResultCode}");
+                    }
+                }
             }
-            else
+            catch (OperationCanceledException)
             {
-                throw new AstarteTransportException
-                ($"Error connecting to MQTT. Code: {result.ResultCode}");
+                Console.WriteLine("Timeout while connecting.");
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs:145]:
```suggestion
                Trace.WriteLine("Timeout while connecting.");
```

File: AstarteDeviceSDKCSharp/Device/AstarteDevice.cs [hunk 2]
```diff
@@ -180,7 +180,19 @@ private bool EventualyReconnect()
...
+                    try
+                    {
+                        Task.Run(async () => await Connect()).Wait(interval);
+                    }
+                    catch (AggregateException ex)
+                    {
+                        foreach (var innerException in ex.InnerExceptions)
+                        {
+                            // Handle each individual exception
+                            Console.WriteLine($"Inner Exception: {innerException.GetType().Name}: {innerException.Message}");
```
Comment (nedimtokic) [inline AstarteDeviceSDKCSharp/Device/AstarteDevice.cs:193]:
```suggestion
                            Trace.WriteLine($"Inner Exception: {innerException.GetType().Name}: {innerException.Message}");
```

**Review by nedimtokic — CHANGES_REQUESTED** [review]

File: AstarteDeviceSDKCSharp/Device/AstarteDevice.cs [hunk 3]
```diff
@@ -180,7 +180,19 @@ private bool EventualyReconnect()
                     {
                         interval = MAX_INCREMENT_INTERVAL;
                     }
-                    Task.Run(async () => await Connect()).Wait(interval);
+
+                    try
+                    {
+                        Task.Run(async () => await Connect()).Wait(interval);
+                    }
+                    catch (AggregateException ex)
```
Comment (harlem88) [inline AstarteDeviceSDKCSharp/Device/AstarteDevice.cs:188]: Why did you catch all exception with `AggregateException` here?

File: AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs [hunk 4]
```diff
@@ -114,18 +121,30 @@ public override async Task Connect()
...
             }
 
-            var result = await _client.ConnectAsync(_connectionInfo.GetMqttConnectOptions(),
-                    CancellationToken.None);
-
-            if (result.ResultCode == MqttClientConnectResultCode.Success)
+            try
             {
-                await CompleteAstarteConnection(result.IsSessionPresent);
+                using (var timeoutToken = new CancellationTokenSource(TimeSpan.FromMilliseconds(500)))
```
Comment (harlem88) [inline AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs:126]: Why did you add a CancellationToken with `500` ms?

File: AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttV1Transport.cs [hunk 5]
```diff
@@ -69,11 +69,20 @@ public AstarteMqttV1Transport(MutualSSLAuthenticationMqttConnectionInfo connecti
 
             try
             {
-                await DoSendMqttMessage(topic, payload, qos);
+                if (_client.TryPingAsync().Result)
```
Comment (harlem88) [inline AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttV1Transport.cs:72]: What happens here if a disconnection occurs immediately after the ping method or during `DoSendMqttMessage` method? I don't believe this approach solves all the issue related to duplicate and lost messages.

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs (hunk 4 above)
Comment (osmanhadzic) [reply AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs:126]: Mqttnet suggestion is put one second, but i think that is too much and I put on 500 ms.
https://github.com/dotnet/MQTTnet/blob/f79649d71b6544952986bac7f645482617a613ba/Samples/Client/Client_Connection_Samples.cs#L68

File: AstarteDeviceSDKCSharp/Device/AstarteDevice.cs (hunk 3 above)
Comment (osmanhadzic) [reply AstarteDeviceSDKCSharp/Device/AstarteDevice.cs:188]: If does not catch this exception, produced mess in threads. Do you have better solution for handling exception? https://github.com/dotnet/MQTTnet/blob/f79649d71b6544952986bac7f645482617a613ba/Samples/Client/Client_Connection_Samples.cs#L68

File: AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttV1Transport.cs (hunk 5 above)
Comment (osmanhadzic) [reply AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttV1Transport.cs:72]: If happens that method `DoSendMqttMessages` throw `MqtttCommunicationException`, exeption handling was here https://github.com/osmanhadzic/astarte-device-sdk-csharp-public/blob/ea7eb4e6ca83ce49f72c43f5ad4bc4f708be0488/AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttV1Transport.cs#L197

File: AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs (hunk 4 above)
Comment (harlem88) [reply AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs:126]: So if the connection method takes more than 500 ms to connect, the task is cancelled?

Comment (osmanhadzic) [reply AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs:126]: If device does not connect in 500 ms Cancellationtoken stop Task.

Comment (harlem88) [reply AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttTransport.cs:126]: The timeout value is too low and cannot be a fixed value, because different networks have different latencies.
I think the best option is to use the same of [MqttClientOptionBuilder]( https://github.com/dotnet/MQTTnet/blob/eb25500f508751365d18ca1e0d7d1b911b1d1c66/Source/MQTTnet/Client/Options/MqttClientOptionsBuilder.cs#L336), leaving the user to modify it via Device constructor.
If it the timeout option is not defined, use the default value of `.net  socket`.

File: AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttV1Transport.cs [hunk 6]
```diff
@@ -167,11 +180,21 @@ public override async Task SendIntrospection()
...
+                {
+                    await DoSendMqttMessage(topic, payload, (MqttQualityOfServiceLevel)qos);
+                }
+                else
+                {
+                    HandleDatastreamFailedPublish(
+                    new MqttCommunicationException("Broker is not avelabe"),
```
Comment (harlem88) [inline AstarteDeviceSDKCSharp/Transport/MQTT/AstarteMqttV1Transport.cs:190]:
```suggestion
                    new MqttCommunicationException("Broker is not available"),
```

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by nedimtokic — APPROVED** [review]

**Review by harlem88 — APPROVED** [review]

**Review by harlem88 — APPROVED** [review]
